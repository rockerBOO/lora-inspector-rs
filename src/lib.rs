use std::fmt::Debug;
// use std::alloc::Global;
use std::string::String;

// use candle_core::safetensors;
// use std::error::Error;
// use std::marker::Send;
// use std::marker::Sync;

mod file;
mod metadata;
mod network;
mod norms;
mod weight;
mod worker;

pub type Result<T> = std::result::Result<T, InspectorError>;

#[derive(Debug)]
pub enum InspectorError {
    Candle(candle_core::Error),
    SafeTensor(safetensors::SafeTensorError),
    Load(String),
    Msg(String),
}

impl InspectorError {
    fn candle(err: candle_core::Error) -> InspectorError {
        InspectorError::Candle(err)
    }

    fn safetensor(err: safetensors::SafeTensorError) -> InspectorError {
        InspectorError::SafeTensor(err)
    }
}

impl From<candle_core::Error> for InspectorError {
    fn from(err: candle_core::Error) -> InspectorError {
        InspectorError::candle(err)
    }
}

impl From<safetensors::SafeTensorError> for InspectorError {
    fn from(err: safetensors::SafeTensorError) -> InspectorError {
        InspectorError::safetensor(err)
    }
}

pub fn get_base_name(name: &str) -> String {
    name.split('.')
        .filter(|part| !matches!(*part, "weight" | "lora_up" | "lora_down" | "alpha"))
        .fold(String::new(), |acc, v| {
            if acc.is_empty() {
                v.to_owned()
            } else {
                format!("{acc}.{v}")
            }
        })
}

#[cfg(test)]
mod tests {
    use crate::get_base_name;

    #[test]
    fn get_base_name_test() {
        let base_name = get_base_name("lora_unet_up_blocks_1_attentions_1_proj_out.lora_up.weight");
        assert_eq!(base_name, "lora_unet_up_blocks_1_attentions_1_proj_out");

        let base_name =
            get_base_name("lora_unet_up_blocks_1_attentions_1_proj_out.lora_down.weight");
        assert_eq!(base_name, "lora_unet_up_blocks_1_attentions_1_proj_out");

        let base_name = get_base_name("lora_unet_up_blocks_1_attentions_1_proj_out.alpha");
        assert_eq!(base_name, "lora_unet_up_blocks_1_attentions_1_proj_out");
    }
}
