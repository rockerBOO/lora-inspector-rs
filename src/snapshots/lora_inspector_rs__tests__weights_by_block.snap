---
source: src/lib.rs
expression: result
---
{
  "unet": {
    "down_02": 2.2030147237448237,
    "down_00": 2.0547019995405065,
    "up_02": 2.1206552568330026,
    "down_01": 2.230538075096385,
    "mid_01": 2.2196846091685765,
    "up_01": 2.239336001916731,
    "mid_00": 2.152647303863596,
    "mid_02": 2.224794629779768,
    "up_00": 2.1742004173330853
  },
  "text_encoder": {
    "layer_08": 1.233122912380843,
    "layer_02": 1.238819093326412,
    "layer_07": 1.229626061262309,
    "layer_04": 1.2323092332639434,
    "layer_01": 1.237123584120902,
    "layer_10": 1.2340809383377243,
    "layer_06": 1.2311805503592266,
    "layer_09": 1.2418230235341936,
    "layer_05": 1.227648301288342,
    "layer_00": 1.2405343164401639,
    "layer_03": 1.232601893524758,
    "layer_11": 1.2385596831422123
  }
}
